// browser-style-dictionary/lib/common/formats.js
/*
 * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

const fs = require('fs');
const path = require('path');
const _template = require('lodash/template');
const GroupMessages = require('../utils/groupMessages');
const { fileHeader, formattedVariables, getTypeScriptType, iconsWithPrefix, minifyDictionary, sortByReference, createPropertyFormatter, sortByName, setSwiftFileProperties } = require('./formatHelpers');

const SASS_MAP_FORMAT_DEPRECATION_WARNINGS = GroupMessages.GROUP.SassMapFormatDeprecationWarnings;

/**
 * @namespace Formats
 */

module.exports = {
  /**
   * Creates a CSS file with variable definitions based on the style dictionary
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @param {string} [options.selector] - Override the root css selector
   * @example
   * ```css
   * :root {
   *   --color-background-base: #f0f0f0;
   *   --color-background-alt: #eeeeee;
   * }
   * ```
   */
  'css/variables': function({dictionary, options={}, file}) {
    const selector = options.selector ? options.selector : `:root`;
    const { outputReferences } = options;
    return fileHeader({file}) +
      `${selector} {\n` +
      formattedVariables({format: 'css', dictionary, outputReferences}) +
      `\n}\n`;
  },

  /**
   * Creates a SCSS file with a flat map based on the style dictionary
   *
   * Name the map by adding a 'mapName' attribute on the file object in your config.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```scss
   * $tokens: (
   *   'color-background-base': #f0f0f0;
   *   'color-background-alt': #eeeeee;
   * )
   * ```
   */
  'scss/map-flat': function({dictionary, options, file}) {
    const template = _template("<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\n%>\n<%= fileHeader({file, commentStyle: 'long'}) %><%\n    var output = '';\n    output += `$${file.mapName||'tokens'}: (\\n`;\n    output += allTokens.map(function(prop){\n        var line = '';\n        if(prop.comment) {\n        line += '  // ' + prop.comment + '\\n';\n        }\n        line += '  \\'' + prop.name + '\\': ' + (prop.attributes.category==='asset' ? '\"'+prop.value+'\"' : prop.value)\n        return line;\n    }).join(',\\n');\n    output += '\\n);';\n    print(output);\n%>\n");
    const { allTokens } = dictionary;
    return template({allTokens, file, options, fileHeader});
  },

  // This will soon be removed, is left here only for backwards compatibility
  'sass/map-flat': function({dictionary, options, file}) {
    GroupMessages.add(SASS_MAP_FORMAT_DEPRECATION_WARNINGS, "sass/map-flat");
    return module.exports['scss/map-flat']({dictionary, options, file});
  },

  /**
   * Creates a SCSS file with a deep map based on the style dictionary.
   *
   * Name the map by adding a 'mapName' attribute on the file object in your config.
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @param {Boolean} [options.themeable=true] - Whether or not tokens should default to being themeable, if not otherwise specified per token.
   * @example
   * ```scss
   * $color-background-base: #f0f0f0 !default;
   * $color-background-alt: #eeeeee !default;
   *
   * $tokens: {
   *   'color': (
   *     'background': (
   *       'base': $color-background-base,
   *       'alt': $color-background-alt
   *     )
   *   )
   * )
   * ```
   */
  'scss/map-deep': function({dictionary, options, file}) {
    const mapTemplate = _template("<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n<%\n    // output the list of tokens as a Sass nested map\n    // (the values are pointing to the variables)\n    //\n    print(`$${file.mapName||'tokens'}: ${processJsonNode(dictionary.properties, 0)};\\n`);\n\n    // recursive function to process a properties JSON node\n    //\n    function processJsonNode(obj, depth) {\n        var output = '';\n        if (obj.hasOwnProperty('value')) {\n            // if we have found a leaf (a property with a value) append the value\n            output += `$${obj.name}`;\n        } else {\n            // if we have found a group of properties, use the Sass group \"(...)\" syntax and loop -recursively- on the children\n            output += '(\\n'\n            output += Object.keys(obj).map(function(newKey) {\n            var newProp = obj[newKey];\n            var indent = '  '.repeat(depth+1);\n            return `${indent}'${newKey}': ${processJsonNode(newProp, depth + 1)}`;\n            }).join(',\\n');\n            output += '\\n' + '  '.repeat(depth) + ')';\n        }\n        return output;\n    }\n%>");

    // Default the "themeable" option to true for backward compatibility.
    const { outputReferences, themeable = true } = options;
    return '\n' + 
      fileHeader({ file, commentStyle: 'long' }) +
      formattedVariables({ format: 'sass', dictionary, outputReferences, themeable })
      + '\n' +
      mapTemplate({dictionary, file});
  },

  // This will soon be removed, is left here only for backwards compatibility
  'sass/map-deep': function({dictionary, options, file}) {
    GroupMessages.add(SASS_MAP_FORMAT_DEPRECATION_WARNINGS, "sass/map-deep");
    return module.exports['scss/map-deep']({dictionary, options, file});
  },

  /**
   * Creates a SCSS file with variable definitions based on the style dictionary.
   *
   * Add `!default` to any variable by setting a `themeable: true` attribute in the token's definition.
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @param {Boolean} [options.themeable=false] - Whether or not tokens should default to being themeable, if not otherwise specified per token.
   * @example
   * ```scss
   * $color-background-base: #f0f0f0;
   * $color-background-alt: #eeeeee !default;
   * ```
   */
  'scss/variables': function({dictionary, options, file}) {
    const { outputReferences, themeable = false } = options;
    return fileHeader({file, commentStyle: 'short'}) +
      formattedVariables({format: 'sass', dictionary, outputReferences, themeable});
  },

  /**
   * Creates a SCSS file with variable definitions and helper classes for icons
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```scss
   * $content-icon-email: '\E001';
   * .icon.email:before { content:$content-icon-email; }
   * ```
   */
  'scss/icons': function({dictionary, options, file}) {
    return fileHeader({file, commentStyle: 'short'}) + iconsWithPrefix('$', dictionary.allTokens, options);
  },

  /**
   * Creates a LESS file with variable definitions based on the style dictionary
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```less
   * \@color-background-base: #f0f0f0;
   * \@color-background-alt: #eeeeee;
   * ```
   */
  'less/variables': function({dictionary, options, file}) {
    const { outputReferences } = options;
    return fileHeader({file, commentStyle: 'short'}) +
      formattedVariables({format: 'less', dictionary, outputReferences});
  },

  /**
   * Creates a LESS file with variable definitions and helper classes for icons
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```less
   * \@content-icon-email: '\E001';
   * .icon.email:before { content:\@content-icon-email; }
   * ```
   */
  'less/icons': function({dictionary, options, file}) {
    return fileHeader({file, commentStyle: 'short'}) + iconsWithPrefix('@', dictionary.allTokens, options);
  },

  /**
   * Creates a Stylus file with variable definitions based on the style dictionary
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```stylus
   * $color-background-base= #f0f0f0;
   * $color-background-alt= #eeeeee;
   * ```
   */
  'stylus/variables': function({dictionary, options, file}) {
    const { outputReferences } = options;
    return fileHeader({file, commentStyle: 'short'}) +
      formattedVariables({format: 'stylus', dictionary, outputReferences});
  },

  /**
   * Creates a CommonJS module with the whole style dictionary
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```js
   * module.exports = {
   *   color: {
   *     base: {
   *        red: {
   *          value: '#ff0000'
   *        }
   *     }
   *   }
   * }
   * ```
   */
  'javascript/module': function({dictionary, file}) {
    return fileHeader({file}) +
    'module.exports = ' +
      JSON.stringify(dictionary.tokens, null, 2) + ';';
  },

  /**
   * Creates a CommonJS module with the whole style dictionary flattened to a single level.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```js
   * module.exports = {
   *  "ColorBaseRed": "#ff0000"
   *}
   *```
   */
  'javascript/module-flat': function({dictionary, file}) {
    return fileHeader({file}) +
      'module.exports = ' +
        module.exports['json/flat']({dictionary}) + ';';
  },

  /**
   * Creates a JS file a global var that is a plain javascript object of the style dictionary.
   * Name the variable by adding a 'name' attribute on the file object in your config.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```js
   * var StyleDictionary = {
   *   color: {
   *     base: {
   *        red: {
   *          value: '#ff0000'
   *        }
   *     }
   *   }
   * }
   * ```
   */
  'javascript/object': function({dictionary, file}) {
    return  fileHeader({file}) +
      'var ' +
      (file.name || '_styleDictionary') +
      ' = ' +
      JSON.stringify(dictionary.tokens, null, 2) +
      ';';
  },

  /**
   * Creates a [UMD](https://github.com/umdjs/umd) module of the style
   * dictionary. Name the module by adding a 'name' attribute on the file object
   * in your config.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```js
   * (function(root, factory) {
   *   if (typeof module === "object" && module.exports) {
   *     module.exports = factory();
   *   } else if (typeof exports === "object") {
   *     exports["_styleDictionary"] = factory();
   *   } else if (typeof define === "function" && define.amd) {
   *     define([], factory);
   *   } else {
   *     root["_styleDictionary"] = factory();
   *   }
   * }(this, function() {
   *   return {
   *     "color": {
   *       "red": {
   *         "value": "#FF0000"
   *       }
   *     }
   *   };
   * }))
   * ```
   */
  'javascript/umd': function({dictionary, file}) {
    var name = file.name || '_styleDictionary'
    return fileHeader({file}) +
      '(function(root, factory) {\n' +
      '  if (typeof module === "object" && module.exports) {\n' +
      '    module.exports = factory();\n' +
      '  } else if (typeof exports === "object") {\n' +
      '    exports["' + name + '"] = factory();\n' +
      '  } else if (typeof define === "function" && define.amd) {\n' +
      '    define([], factory);\n' +
      '  } else {\n' +
      '    root["' + name + '"] = factory();\n' +
      '  }\n' +
      '}(this, function() {\n' +
      '  return ' + JSON.stringify(dictionary.tokens, null, 2) + ';\n' +
      '}))\n'
  },

  /**
   * Creates a ES6 module of the style dictionary.
   *
   * ```json
   * {
   *   "platforms": {
   *     "js": {
   *       "transformGroup": "js",
   *       "files": [
   *         {
   *           "format": "javascript/es6",
   *           "destination": "colors.js",
   *           "filter": {
   *             "attributes": {
   *               "category": "color"
   *             }
   *           }
   *         }
   *       ]
   *     }
   *   }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```js
   * export const ColorBackgroundBase = '#ffffff';
   * export const ColorBackgroundAlt = '#fcfcfcfc';
   * ```
   */
  'javascript/es6': function({dictionary, file}) {
    return fileHeader({file}) +
      dictionary.allTokens.map(function(token) {
        var to_ret = 'export const ' + token.name + ' = ' + JSON.stringify(token.value) + ';';
        if (token.comment)
          to_ret = to_ret.concat(' // ' + token.comment);
        return to_ret;
      }).join('\n');
  },

  // TypeScript declarations
  /**
   * Creates TypeScript declarations for ES6 modules
   *
   * ```json
   * {
   *   "platforms": {
   *     "ts": {
   *       "transformGroup": "js",
   *       "files": [
   *         {
   *           "format": "javascript/es6",
   *           "destination": "colors.js"
   *         },
   *         {
   *           "format": "typescript/es6-declarations",
   *           "destination": "colors.d.ts"
   *         }
   *       ]
   *     }
   *   }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```typescript
   * export const ColorBackgroundBase : string;
   * export const ColorBackgroundAlt : string;
   * ```
   */
  'typescript/es6-declarations': function({dictionary, file}) {
    return fileHeader({file}) +
      dictionary.allProperties.map(function(prop) {
        var to_ret_prop = 'export const ' + prop.name + ' : ' + getTypeScriptType(prop.value) + ';';
        if (prop.comment)
          to_ret_prop = to_ret_prop.concat(' // ' + prop.comment);
        return to_ret_prop;
      }).join('\n');
  },

  /**
   * Creates TypeScript declarations for CommonJS module
   *
   * ```json
   * {
   *   "platforms": {
   *     "ts": {
   *       "transformGroup": "js",
   *       "files": [
   *         {
   *           "format": "javascript/module",
   *           "destination": "colors.js"
   *         },
   *         {
   *           "format": "typescript/module-declarations",
   *           "destination": "colors.d.ts"
   *         }
   *       ]
   *     }
   *   }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```typescript
   * export default tokens;
   * declare interface DesignToken { value: string; name?: string; path?: string[]; comment?: string; attributes?: any; original?: any; }
   * declare const tokens: {
   *  "color": {
   *    "red": DesignToken
   *  }
   * }
   * ```
   *
   * As you can see above example output this does not generate 100% accurate d.ts.
   * This is a compromise between of what style-dictionary can do to help and not bloating the library with rarely used dependencies.
   *
   * Thankfully you can extend style-dictionary very easily:
   *
   * ```js
   * const JsonToTS = require('json-to-ts');
   * StyleDictionaryPackage.registerFormat({
   *   name: 'typescript/accurate-module-declarations',
   *   formatter: function({ dictionary }) {
   *     return 'declare const root: RootObject\n' +
   *     'export default root\n' +
   *     JsonToTS(dictionary.properties).join('\n');
   *   },
   * });
   * ```
   */
  'typescript/module-declarations': function({dictionary, file, options}) {
    const {moduleName=`tokens`} = options;
    function treeWalker(obj) {
      let type = Object.create(null);
      let has = Object.prototype.hasOwnProperty.bind(obj);
      if (has('value')) {
        type = 'DesignToken';
      } else {
        for (var k in obj) if (has(k)) {
          switch (typeof obj[k]) {
            case 'object':
              type[k] = treeWalker(obj[k]);
          }
        }
      }
      return type;
    }
    const designTokenInterface = "/*\n * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n *\n * Licensed under the Apache License, Version 2.0 (the \"License\"). You may not use this file except in compliance with\n * the License. A copy of the License is located at\n *\n * http://www.apache.org/licenses/LICENSE-2.0\n *\n * or in the \"license\" file accompanying this file. This file is distributed on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR\n * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions\n * and limitations under the License.\n */\n\n/**\n * This type is also used in the `typescript/module-declarations` format,\n * but we only want to include the lines of the DesignToken interface.\n * The comments surrounding the definition are used to trim the lines of\n * this file in that format.\n */\n\n//start\ninterface DesignToken {\n  value: any;\n  name?: string;\n  comment?: string;\n  themeable?: boolean;\n  attributes?: {\n    category?: string;\n    type?: string;\n    item?: string;\n    subitem?: string;\n    state?: string;\n    [key: string]: any;\n  };\n  [key: string]: any;\n}\n//end\n\nexport { DesignToken };\nexport interface DesignTokens {\n  [key: string]: DesignTokens | DesignToken;\n}"

    // get the first and last lines to add to the format by
    // looking for the first and last single-line comment
    const lines = designTokenInterface
      .split('\n');
    const firstLine = lines.indexOf(`//start`) + 1;
    const lastLine = lines.indexOf(`//end`);

    const output = fileHeader({file}) +
`export default ${moduleName};

declare ${lines.slice(firstLine, lastLine).join(`\n`)}

declare const ${moduleName}: ${JSON.stringify(treeWalker(dictionary.tokens), null, 2)}`;

    // JSON stringify will quote strings, because this is a type we need
    // it unquoted.
    return output.replace(/"DesignToken"/g,'DesignToken');
  },

  // Android templates
  /**
   * Creates a [resource](https://developer.android.com/guide/topics/resources/providing-resources) xml file. It is recommended to use a filter with this format
   * as it is generally best practice in Android development to have resource files
   * organized by type (color, dimension, string, etc.). However, a resource file
   * with mixed resources will still work.
   *
   * This format will try to use the proper resource type for each token based on
   * the category (color => color, size => dimen, etc.). However if you want to
   * force a particular resource type you can provide a 'resourceType' attribute
   * on the file configuration. You can also provide a 'resourceMap' if you
   * don't use Style Dictionary's built-in CTI structure.
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```xml
   * <?xml version="1.0" encoding="UTF-8"?>
   * <resources>
   *  <color name="color_base_red_5">#fffaf3f2</color>
   *  <color name="color_base_red_30">#fff0cccc</color>
   *  <dimen name="size_font_base">14sp</color>
   * ```
   */
  'android/resources': function({dictionary, options, file}) {
    const template = _template(
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar resourceType = file.resourceType || null;\n\nvar resourceMap = file.resourceMap || {\n  size: 'dimen',\n  color: 'color',\n  string: 'string',\n  content: 'string',\n  time: 'integer',\n  number: 'integer'\n};\n\nfunction propToType(prop) {\n  if (resourceType) {\n    return resourceType;\n  }\n  if (resourceMap[prop.attributes.category]) {\n    return resourceMap[prop.attributes.category];\n  }\n  return 'string';\n}\n\nfunction propToValue(prop) {\n  if (file.options && file.options.outputReferences && dictionary.usesReference(prop.original.value)) {\n    return `@${propToType(prop)}/${dictionary.getReferences(prop.original.value)[0].name}`;\n  } else {\n    return prop.value;\n  }\n} %>\n<%= fileHeader({file, commentStyle: 'xml'}) %>\n<resources>\n  <% dictionary.allTokens.forEach(function(prop) {\n  %><<%= propToType(prop) %> name=\"<%= prop.name %>\"><%= propToValue(prop) %></<%= propToType(prop) %>><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %>\n  <% }); %>\n</resources>\n"
    );
    return template({dictionary, file, options, fileHeader});
  },

  /**
   * Creates a color resource xml file with all the colors in your style dictionary.
   *
   * It is recommended to use the 'android/resources' format with a custom filter
   * instead of this format:
   *
   * ```javascript
   * format: 'android/resources',
   * filter: {
   *   attributes: { category: 'color' }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```xml
   * <?xml version="1.0" encoding="UTF-8"?>
   * <resources>
   *  <color name="color_base_red_5">#fffaf3f2</color>
   *  <color name="color_base_red_30">#fff0cccc</color>
   *  <color name="color_base_red_60">#ffe19d9c</color>
   * ```
   */
  'android/colors': function({dictionary, options, file}) {
    const template = _template(
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar props = dictionary.allTokens.filter(function(prop) {\n return prop.attributes.category === 'color';\n});\n%>\n<%= fileHeader({file, commentStyle: 'xml'}) %>\n<resources>\n  <% props.forEach(function(prop) {\n  %><color name=\"<%= prop.name %>\"><%= prop.value %></color><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %>\n  <% }); %>\n</resources>\n"
    );
    return template({dictionary, file, options, fileHeader});
  },

  /**
   * Creates a dimen resource xml file with all the sizes in your style dictionary.
   *
   * It is recommended to use the 'android/resources' format with a custom filter
   * instead of this format:
   *
   * ```javascript
   * format: 'android/resources',
   * filter: {
   *   attributes: { category: 'size' }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```xml
   * <?xml version="1.0" encoding="UTF-8"?>
   * <resources>
   *  <dimen name="size_padding_tiny">5.00dp</dimen>
   *  <dimen name="size_padding_small">10.00dp</dimen>
   *  <dimen name="size_padding_medium">15.00dp</dimen>
   * ```
   */
  'android/dimens': function({dictionary, options, file}) {
    const template = _template(
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar props = dictionary.allTokens.filter(function(prop) {\n  return prop.attributes.category === 'size' && prop.attributes.type !== 'font'\n}); %>\n<%= fileHeader({file, commentStyle: 'xml'}) %>\n<resources>\n  <% props.forEach(function(prop) {\n  %><dimen name=\"<%= prop.name %>\"><%= prop.value %></dimen><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %>\n  <% }); %>\n</resources>\n"
    );
    return template({dictionary, file, options, fileHeader});
  },

  /**
   * Creates a dimen resource xml file with all the font sizes in your style dictionary.
   *
   * It is recommended to use the 'android/resources' format with a custom filter
   * instead of this format:
   *
   * ```javascript
   * format: 'android/resources',
   * filter: {
   *   attributes: { category: 'size' }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```xml
   * <?xml version="1.0" encoding="UTF-8"?>
   * <resources>
   *  <dimen name="size_font_tiny">10.00sp</dimen>
   *  <dimen name="size_font_small">13.00sp</dimen>
   *  <dimen name="size_font_medium">15.00sp</dimen>
   * ```
   */
  'android/fontDimens': function({dictionary, options, file}) {
    const template = _template(
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar props = dictionary.allTokens.filter(function(prop) {\n  return prop.attributes.category === 'size' && prop.attributes.type === 'font';\n}); %>\n<%= fileHeader({file, commentStyle: 'xml'}) %>\n<resources>\n  <% props.forEach(function(prop) {\n  %><dimen name=\"<%= prop.name %>\"><%= prop.value %></dimen><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %>\n  <% }); %>\n</resources>\n"
    );
    return template({dictionary, file, options, fileHeader});
  },

  /**
   * Creates a resource xml file with all the integers in your style dictionary. It filters your
   * design tokens by `token.attributes.category === 'time'`
   *
   * It is recommended to use the 'android/resources' format with a custom filter
   * instead of this format:
   *
   * ```javascript
   * format: 'android/resources',
   * filter: {
   *   attributes: { category: 'time' }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @todo Update the filter on this.
   * @example
   * ```xml
   * <?xml version="1.0" encoding="UTF-8"?>
   * <resources>
   *   <integer name="time_duration_short">1000</integer>
   *   <integer name="time_duration_medium">2000</integer>
   *   <integer name="time_duration_long">4000</integer>
   * ```
   */
  'android/integers': function({dictionary, options, file}) {
    const template = _template(
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar props = dictionary.allTokens.filter(function(prop) {\n  return prop.attributes.category === 'time';\n}); %>\n<%= fileHeader({file, commentStyle: 'xml'}) %>\n<resources>\n  <% props.forEach(function(prop) {\n  %><integer name=\"<%= prop.name %>\"><%= prop.value %></integer><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %>\n  <% }); %>\n</resources>\n"
    );
    return template({dictionary, file, options, fileHeader});
  },

  /**
   * Creates a resource xml file with all the strings in your style dictionary. Filters your
   * design tokens by `token.attributes.category === 'content'`
   *
   * It is recommended to use the 'android/resources' format with a custom filter
   * instead of this format:
   *
   * ```javascript
   * format: 'android/resources',
   * filter: {
   *   attributes: { category: 'content' }
   * }
   * ```
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```xml
   * <?xml version="1.0" encoding="UTF-8"?>
   * <resources>
   *   <string name="content_icon_email">&#xE001;</string>
   *   <string name="content_icon_chevron_down">&#xE002;</string>
   *   <string name="content_icon_chevron_up">&#xE003;</string>
   * ```
   */
  'android/strings': function({dictionary, options, file}) {
    const template = _template(
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar props = dictionary.allTokens.filter(function(prop) {\n  return prop.attributes.category === 'content';\n}); %>\n<%= fileHeader({file, commentStyle: 'xml'}) %>\n<resources>\n  <% props.forEach(function(prop) {\n  %><string name=\"<%= prop.name %>\"><%= prop.value %></string><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %>\n  <% }); %>\n</resources>\n"
    );
    return template({dictionary, file, options, fileHeader});
  },

  // Compose templates
  /**
   * Creates a Kotlin file for Compose containing an object with a `val` for each property.
   *
   * @memberof Formats
   * @kind member
   * @param {String} className The name of the generated Kotlin object
   * @param {String} packageName The package for the generated Kotlin object
   * @param {Object} options
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```kotlin
   * package com.example.tokens;
   *
   * import androidx.compose.ui.graphics.Color
   *
   * object StyleDictionary {
   *  val colorBaseRed5 = Color(0xFFFAF3F2)
   * }
   * ```
   */
  'compose/object': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2020 Target Corporation. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n<%= fileHeader({file, commentStyle: 'short'}) %>\n\npackage <%= file.packageName %>;\n\n<%  // Ideally we would selectively import the correct classes & functions based on whether the applied transforms actually need them. %>\nimport androidx.compose.ui.graphics.Color\nimport androidx.compose.ui.unit.*\n\nobject <%= file.className %> {\n  <%= allProperties.map(function(prop) {\n    let comment = \"\";\n    if (prop.comment) {\n      comment = `/** ${prop.comment} */\\n  `\n    }\n    return comment + 'val ' + formatProperty(prop)\n  }).join('\\n  ' ) %>\n}\n"
    );
    let allProperties;
    const { outputReferences } = options;
    const formatProperty = createPropertyFormatter({
      outputReferences,
      dictionary,
      formatting: {
        suffix: '',
        commentStyle: 'none' // We will add the comment in the format template
      }
    });

    if (outputReferences) {
      allProperties = [...dictionary.allProperties].sort(sortByReference(dictionary));
    } else {
      allProperties = [...dictionary.allProperties].sort(sortByName);
    }

    return template({allProperties, file, options, formatProperty, fileHeader});
  },

  // iOS templates

  /**
   * Creates an Objective-C header file with macros for design tokens
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```objectivec
   * #import <Foundation/Foundation.h>
   * #import <UIKit/UIKit.h>
   *
   * #define ColorFontLink [UIColor colorWithRed:0.00f green:0.47f blue:0.80f alpha:1.00f]
   * #define SizeFontTiny 176.00f
   * ```
   */
  'ios/macros': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n#import <Foundation/Foundation.h>\n#import <UIKit/UIKit.h>\n\n<% dictionary.allTokens.forEach(function(prop) {\n  %>#define <%= prop.name %> <%= prop.value %>\n<% }); %>\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C plist file
   *
   * @memberof Formats
   * @kind member
   * @todo Fix this template and add example and usage
   */
  'ios/plist': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n\nvar props = dictionary.allTokens.filter(function(prop) {\n  return prop.attributes.category !== 'asset' &&\n         prop.attributes.category !== 'border' &&\n         prop.attributes.category !== 'shadow' &&\n         prop.attributes.category !== 'transition';\n}); %>\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<%= fileHeader({ file, commentStyle: 'xml' }) %>\n<plist version=\"1.0\">\n  <dict>\n    <% props.forEach(function(prop) {\n    %><key><%= prop.name %></key>\n    <% if (prop.attributes.category === 'color') { %><dict>\n      <key>r</key>\n      <real><%= prop.value[0]/255 %></real>\n      <key>g</key>\n      <real><%= prop.value[1]/255 %></real>\n      <key>b</key>\n      <real><%= prop.value[2]/255 %></real>\n      <key>a</key>\n      <real>1</real>\n      </dict>\n    <% } else if (prop.attributes.category === 'size') { %></dict>\n      <integer><%= prop.value %></integer>\n    <% } else { %></dict>\n      <string><%= prop.value %></string>\n    <% } %><% if (prop.comment) { %><!-- <%= prop.comment %> --><% } %><% }); %>\n  </dict>\n</plist>\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C implementation file of a style dictionary singleton class
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/singleton.m': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({ file, commentStyle: 'short' })%>\n#import \"<%= file.className %>.h\"\n\n@implementation <%= file.className %>\n\n+ (NSDictionary *)getProperty:(NSString *)keyPath {\n  return [[self properties] valueForKeyPath:keyPath];\n}\n\n+ (nonnull)getValue:(NSString *)keyPath {\n  return [[self properties] valueForKeyPath:[NSString stringWithFormat:@\"%@.value\", keyPath]];\n}\n\n+ (NSDictionary *)properties {\n  static NSDictionary * dictionary;\n  static dispatch_once_t onceToken;\n\n  dispatch_once(&onceToken, ^{\n    dictionary = <%= buildDictionary(dictionary.properties) %>;\n  });\n\n  return dictionary;\n}\n\n@end\n\n<% function buildDictionary(props, indent) {\n  indent = indent || '  ';\n  var to_ret = '@{\\n';\n  if (props.hasOwnProperty('value')) {\n    var value = props.attributes.category === 'size' || props.attributes.category === 'time' ? '@' + props.value : props.value;\n    to_ret += indent + '@\"value\": ' + value + ',\\n';\n    to_ret += indent + '@\"name\": @\"' + props.name + '\",\\n';\n\n    for(var name in props.attributes) {\n      if (props.attributes[name]) {\n        to_ret += indent + '@\"' + name + '\": @\"' + props.attributes[name] + '\",\\n';\n      }\n    }\n\n    // remove last comma\n    return to_ret.slice(0, -2) + '\\n' + indent + '}';\n  } else {\n    for(var name in props) {\n      to_ret += indent + '@\"' + name + '\": ' + buildDictionary(props[name], indent + '  ') + ',\\n';\n    }\n    // remove last comma\n    return to_ret.slice(0, -2) + '\\n' + indent + '}';\n  }\n} %>\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C header file of a style dictionary singleton class
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/singleton.h': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({ file, commentStyle: 'short' })%>\n#import <Foundation/Foundation.h>\n#import <UIKit/UIKit.h>\n\n@interface <%= file.className %> : NSObject\n\n+ (NSDictionary *)properties;\n+ (NSDictionary *)getProperty:(NSString *)keyPath;\n+ (nonnull)getValue:(NSString *)keyPath;\n\n@end\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C header file of a static style dictionary class
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/static.h': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n//\n%>\n// <%= file.destination %>\n//\n<%= fileHeader({ file, commentStyle: 'short' })%>\n#import <Foundation/Foundation.h>\n\n<% dictionary.allTokens.forEach(function(prop) {  %>\nextern <%= file.type %> const <%= prop.name %>;<% }); %>\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C implementation file of a static style dictionary class
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/static.m': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({ file, commentStyle: 'short' }) %>\n#import \"<%= file.className %>.h\"\n\n<% dictionary.allTokens.forEach(function(prop) {  %>\n<%= file.type %> const <%= prop.name %> = <%= prop.value %>;<% }); %>\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C header file of a color class
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/colors.h': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n#import <UIKit/UIKit.h>\n\ntypedef NS_ENUM(NSInteger, <%= file.type %>) {\n<%= dictionary.allTokens.map(function(prop) { return prop.name; }).join(',\\n') %>\n};\n\n@interface <%= file.className %> : NSObject\n+ (NSArray *)values;\n+ (UIColor *)color:(<%= file.type %>)color;\n@end\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C implementation file of a color class
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/colors.m': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n#import \"<%= file.className %>.h\"\n\n@implementation <%= file.className %>\n\n+ (UIColor *)color:(<%= file.type %>)colorEnum{\n  return [[self values] objectAtIndex:colorEnum];\n}\n\n+ (NSArray *)values {\n  static NSArray* colorArray;\n  static dispatch_once_t onceToken;\n\n  dispatch_once(&onceToken, ^{\n    colorArray = @[\n<%= dictionary.allTokens.map(function(prop){ return prop.value; }).join(',\\n') %>\n    ];\n  });\n\n  return colorArray;\n}\n\n@end\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C header file of strings
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/strings.h': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({ file, commentStyle: 'short' })%>\n#import <Foundation/Foundation.h>\n\n<% dictionary.allTokens.forEach(function(prop) {  %>\nextern NSString * const <%= prop.name %>;<% }); %>\n\n@interface <%= file.className %> : NSObject\n+ (NSArray *)values;\n@end\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates an Objective-C implementation file of strings
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'ios/strings.m': function({dictionary, options, file}) {
    const template = _template(
      "<%\n//\n// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({ file, commentStyle: 'short' }) %>\n#import \"<%= file.className %>.h\"\n\n<% dictionary.allTokens.forEach(function(prop) {  %>\nNSString * const <%= prop.name %> = <%= prop.value %>;<% }); %>\n\n@implementation <%= file.className %>\n\n+ (NSArray *)values {\n  static NSArray* array;\n  static dispatch_once_t onceToken;\n\n  dispatch_once(&onceToken, ^{\n    array = @[\n      <%= dictionary.allTokens.map(buildProperty).join(',\\n') %>\n    ];\n  });\n\n  return array;\n}\n\n@end\n\n<% function buildProperty(prop) {\n  var to_ret = '@{\\n';\n  to_ret += '  ' + '@\"value\": ' + prop.value + ',\\n';\n  to_ret += '  ' + '@\"name\": @\"' + prop.name + '\",\\n';\n\n  for(var name in prop.attributes) {\n    if (prop.attributes[name]) {\n      to_ret += '    ' + '@\"' + name + '\": @\"' + prop.attributes[name] + '\",\\n';\n    }\n  }\n\n  // remove last comma\n  return to_ret.slice(0, -2) + '\\n' + '  ' + '}';\n} %>\n"
    );

    return template({ dictionary, options, file, fileHeader });
  },

  /**
   * Creates a Swift implementation file of a class with values. It adds default `class` object type, `public` access control and `UIKit` import.
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {String} [options.accessControl=public] - Level of [access](https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html) of the generated swift object
   * @param {[String]} [options.import=UIKit] - Modules to import. Can be a string or array of string
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```swift
   * public class StyleDictionary {
   *   public static let colorBackgroundDanger = UIColor(red: 1.000, green: 0.918, blue: 0.914, alpha: 1)
   * }
   * ```
   */
  'ios-swift/class.swift': function({dictionary, options, file, platform}) {
    const template = _template(
      "<%\n//\n// Copyright 2019 Alaska Air Group or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n<%= options.import.map(function(item) {\n    return 'import ' + item\n  }).join('\\n')\n%>\n\n<%= options.accessControl %><%= options.objectType %> <%= file.className %> {\n    <%= allTokens.map(function(prop) {\n      return options.accessControl + 'static let ' + formatProperty(prop);\n    }).join('\\n    ') %>\n}\n"
    );
    let allTokens;
    const { outputReferences } = options;
    options = setSwiftFileProperties(options, 'class', platform.transformGroup);
    const formatProperty = createPropertyFormatter({
      outputReferences,
      dictionary,
      formatting: {
        suffix: ''
      }
    });

    if (outputReferences) {
      allTokens = [...dictionary.allTokens].sort(sortByReference(dictionary));
    } else {
      allTokens = [...dictionary.allTokens].sort(sortByName);
    }

    return template({allTokens, file, options, formatProperty, fileHeader});
  },

  /**
   * Creates a Swift implementation file of an enum with values. It adds default `enum` object type, `public` access control and `UIKit` import.
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {String} [options.accessControl=public] - Level of [access](https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html) of the generated swift object
   * @param {[String]} [options.import=UIKit] - Modules to import. Can be a string or array of string
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```swift
   * public enum StyleDictionary {
   *   public static let colorBackgroundDanger = UIColor(red: 1.000, green: 0.918, blue: 0.914, alpha: 1)
   * }
   * ```
   */
  'ios-swift/enum.swift': function({dictionary, options, file, platform}) {
    const template = _template(
      "<%\n//\n// Copyright 2019 Alaska Air Group or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n<%= options.import.map(function(item) {\n    return 'import ' + item\n  }).join('\\n')\n%>\n\n<%= options.accessControl %><%= options.objectType %> <%= file.className %> {\n    <%= allTokens.map(function(prop) {\n      return options.accessControl + 'static let ' + formatProperty(prop);\n    }).join('\\n    ') %>\n}\n"
    );
    let allTokens;
    const { outputReferences } = options;
    options = setSwiftFileProperties(options, 'enum', platform.transformGroup);
    const formatProperty = createPropertyFormatter({
      outputReferences,
      dictionary,
      formatting: {
        suffix: ''
      }
    });

    if (outputReferences) {
      allTokens = [...dictionary.allTokens].sort(sortByReference(dictionary));
    } else {
      allTokens = [...dictionary.allTokens].sort(sortByName);
    }
    return template({allTokens, file, options, formatProperty, fileHeader});
  },

  /**
   * Creates a Swift implementation file of any given type with values. It has by default `class` object type, `public` access control and `UIKit` import.
   * 
   * ```javascript
   * format: 'ios-swift/any.swift',
   * import: ['UIKit', 'AnotherModule'],
   * objectType: 'struct',
   * accessControl: 'internal',
   * ```
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {String} [options.accessControl=public] - Level of [access](https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html) of the generated swift object
   * @param {[String]} [options.import=UIKit] - Modules to import. Can be a string or array of strings
   * @param {String} [options.objectType=class] - The type of the generated Swift object
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```swift
   * import UIKit
   * import AnotherModule
   * 
   * internal struct StyleDictionary {
   *   internal static let colorBackgroundDanger = UIColor(red: 1.000, green: 0.918, blue: 0.914, alpha: 1)
   * }
   * ```
   */
   'ios-swift/any.swift': function({ dictionary, options, file, platform }) {
    const template = _template(
      "<%\n//\n// Copyright 2019 Alaska Air Group or its affiliates. All Rights Reserved.\n//\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n<%= options.import.map(function(item) {\n    return 'import ' + item\n  }).join('\\n')\n%>\n\n<%= options.accessControl %><%= options.objectType %> <%= file.className %> {\n    <%= allTokens.map(function(prop) {\n      return options.accessControl + 'static let ' + formatProperty(prop);\n    }).join('\\n    ') %>\n}\n"
    );
    let allTokens;
    const { outputReferences } = options;
    options = setSwiftFileProperties(options, options.objectType, platform.transformGroup);
    const formatProperty = createPropertyFormatter({
      outputReferences,
      dictionary,
      formatting: {
        suffix: ''
      }
    });

    if (outputReferences) {
      allTokens = [...dictionary.allTokens].sort(sortByReference(dictionary));
    } else {
      allTokens = [...dictionary.allTokens].sort(sortByName);
    }
    return template({allTokens, file, options, formatProperty, fileHeader});
  },

  // Css templates

  /**
   * Creates CSS file with @font-face declarations
   *
   * @memberof Formats
   * @kind member
   * @todo Add example and usage
   */
  'css/fonts.css': _template(
    "<% Object.values(properties && properties.asset && properties.asset.font || {}).forEach(function(font) {\n  var fileFormatArr = [];\n  if (font.eot) {\n  \tfileFormatArr.push(\"url('../\" + font.eot.value + \"');\\n\\tsrc: url('../\" + font.eot.value + \"?#iefix') format('embedded-opentype')\");\n  }\n  if (font.otf) {\n  \tfileFormatArr.push(\"url('../\" + font.otf.value + \"') format('otf')\");\n  }\n  if (font.ttf) {\n  \tfileFormatArr.push(\"url('../\" + font.ttf.value + \"') format('truetype')\");\n  }\n  if (font.svg) {\n  \tfileFormatArr.push(\"url('../\" + font.svg.value + \"#\" + font.name.value.replace(/ /g,'%20') + \"') format('svg')\");\n  }\n  if (font.woff) {\n  \tfileFormatArr.push(\"url('../\" + font.woff.value + \"') format('woff')\");\n  }\n  if (font.woff2) {\n  \tfileFormatArr.push(\"url('../\" + font.woff2.value + \"') format('woff2')\");\n  }\n%>\n@font-face {\n  font-family: \"<%= font.name.value %>\";\n  src: <%= fileFormatArr.join(',\\n\\t\\t') %>;<% if (font.style) { %>\n  font-style: <%= font.style.value %>;<% } if (font.weight) { %>\n  font-weight: <%= font.weight.value %>;<% } %>\n}\n<% }); %>\n"
  ),

  // Web templates

  /**
   * Creates a JSON file of the style dictionary.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```json
   * {
   *   "color": {
   *     "base": {
   *        "red": {
   *          "value": "#ff0000"
   *        }
   *     }
   *   }
   * }
   * ```
   */
  'json': function({dictionary}) {
    return JSON.stringify(dictionary.tokens, null, 2);
  },

  /**
   * Creates a JSON file of the assets defined in the style dictionary.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```js
   * {
   *   "asset": {
   *     "image": {
   *        "logo": {
   *          "value": "assets/logo.png"
   *        }
   *     }
   *   }
   * }
   * ```
   */
  'json/asset': function({dictionary}) {
    return JSON.stringify({asset: dictionary.tokens.asset}, null, 2);
  },

  /**
   * Creates a JSON nested file of the style dictionary.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```json
   * {
   *   "color": {
   *     "base": {
   *        "red": "#ff0000"
   *     }
   *   }
   * }
   * ```
   */
  'json/nested': function({dictionary}) {
    return JSON.stringify(minifyDictionary(dictionary.tokens), null, 2);
  },

  /**
   * Creates a JSON flat file of the style dictionary.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```json
   * {
   *   "color-base-red": "#ff0000"
   * }
   * ```
   */
  'json/flat': function({dictionary}) {
    return '{\n' + dictionary.allTokens.map(function(token) {
        return `  "${token.name}": ${JSON.stringify(token.value)}`;
      }).join(',\n') + '\n}';
  },

  /**
   * Creates a sketchpalette file of all the base colors
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```json
   * {
   *   "compatibleVersion": "1.0",
   *   "pluginVersion": "1.1",
   *   "colors": [
   *     "#ffffff",
   *     "#ff0000",
   *     "#fcfcfc"
   *   ]
   * }
   * ```
   */
  'sketch/palette': function({dictionary}) {
    var to_ret = {
      'compatibleVersion':'1.0',
      'pluginVersion':'1.1'
    };
    to_ret.colors = dictionary.allTokens
      .filter(function(token) {
        return token.attributes.category === 'color' && token.attributes.type === 'base';
      })
      .map(function(token) {
        return token.value;
      });
    return JSON.stringify(to_ret, null, 2);
  },

  /**
   * Creates a sketchpalette file compatible with version 2 of
   * the sketchpalette plugin. To use this you should use the
   * 'color/sketch' transform to get the correct value for the colors.
   *
   * @memberof Formats
   * @kind member
   * @example
   * ```json
   * {
   *   "compatibleVersion": "2.0",
   *   "pluginVersion": "2.2",
   *   "colors": [
   *     {name: "red", r: 1.0, g: 0.0, b: 0.0, a: 1.0},
   *     {name: "green", r: 0.0, g: 1.0, b: 0.0, a: 1.0},
   *     {name: "blue", r: 0.0, g: 0.0, b: 1.0, a: 1.0}
   *   ]
   * }
   * ```
   */
  'sketch/palette/v2': function({dictionary}) {
    var to_ret = {
      compatibleVersion: '2.0',
      pluginVersion: '2.2',
      colors: dictionary.allTokens.map(function(token) {
        // Merging the token's value, which should be an object with r,g,b,a channels
        return Object.assign({
          name: token.name
        }, token.value)
      })
    };
    return JSON.stringify(to_ret, null, 2);
  },


  // Flutter templates
  /**
   *  Creates a Dart implementation file of a class with values
   *
   * @memberof Formats
   * @kind member
   * @param {Object} options
   * @param {Boolean} [options.showFileHeader=true] - Whether or not to include a comment that has the build date
   * @param {Boolean} [options.outputReferences=false] - Whether or not to keep [references](/#/formats?id=references-in-output-files) (a -> b -> c) in the output.
   * @example
   * ```dart
   * import 'package:flutter/material.dart';
   *
   * class StyleDictionary {
   *   StyleDictionary._();
   *
   *     static const colorBrandPrimary = Color(0x00ff5fff);
   *     static const sizeFontSizeMedium = 16.00;
   *     static const contentFontFamily1 = "NewJune";
   * ```
   */
  'flutter/class.dart': function({dictionary, options, file}) {
    const template = _template(
      "<%\n// Licensed under the Apache License, Version 2.0 (the \"License\").\n// You may not use this file except in compliance with the License.\n// A copy of the License is located at\n//\n// http://www.apache.org/licenses/LICENSE-2.0\n//\n// or in the \"license\" file accompanying this file. This file is distributed\n// on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either\n// express or implied. See the License for the specific language governing\n// permissions and limitations under the License.\n%>\n//\n// <%= file.destination %>\n//\n<%= fileHeader({file, commentStyle: 'short'}) %>\n\nimport 'dart:ui';\n\nclass <%= file.className %> {\n  <%= file.className %>._();\n\n    <%= allTokens.map(function(prop) {\n      return 'static const ' + formatProperty(prop);\n    }).join('\\n    ') %>\n}"
    );
    let allTokens;
    const { outputReferences } = options;
    const formatProperty = createPropertyFormatter({
      outputReferences,
      dictionary
    });

    if (outputReferences) {
      allTokens = [...dictionary.allTokens].sort(sortByReference(dictionary));
    } else {
      allTokens = [...dictionary.allTokens].sort(sortByName)
    }
    return template({allTokens, file, options, formatProperty, fileHeader});
  },
};

// Mark which formats are nested
module.exports['json/nested'].nested = true;
module.exports['javascript/module'].nested = true;
module.exports['javascript/object'].nested = true;
